tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/ForwardGetMethod.scala

```diff
@@ -0,0 +1,136 @@
+    val requestedProperties: Properties = request.properties.getOrElse(Forwards.allProperties)
+    (requestedProperties -- Forwards.allProperties match {
+      case validProperties if validProperties.isEmpty() => getForwards(request, mailboxSession)
+        .reduce(ForwardGetResult.empty)(ForwardGetResult.merge)
+        .map(forwardResult => forwardResult.asResponse(request.accountId))
+        .map(forwardGetResponse => Invocation(
+          methodName = methodName,
+          arguments = Arguments(serializer.serialize(forwardGetResponse, requestedProperties).as[JsObject]),
+          methodCallId = invocation.invocation.methodCallId))
+      case invalidProperties: Properties => SMono.just(Invocation.error(errorCode = ErrorCode.InvalidArguments,
+        description = s"The following properties [${invalidProperties.format}] do not exist.",
+        methodCallId = invocation.invocation.methodCallId))
+    }).map(InvocationWithContext(_, invocation.processingContext))
+      .onErrorResume{ case e: Exception => handleRequestValidationErrors(e, invocation.invocation.methodCallId)
+        .map(errorInvocation => InvocationWithContext(errorInvocation,  invocation.processingContext))}
+  }
+
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, ForwardGetRequest] =
+    serializer.deserializeForwardGetRequest(invocation.arguments.value) match {
+      case JsSuccess(forwardGetRequest, _) => Right(forwardGetRequest)
+      case errors: JsError => Left(new IllegalArgumentException(ResponseSerializer.serialize(errors).toString))
+    }
+
+  private def handleRequestValidationErrors(exception: Exception, methodCallId: MethodCallId): SMono[Invocation] = exception match {
+    case _: MissingCapabilityException => SMono.just(Invocation.error(ErrorCode.UnknownMethod, methodCallId))
+    case e: IllegalArgumentException => SMono.just(Invocation.error(ErrorCode.InvalidArguments, e.getMessage, methodCallId))
+  }
+
+  private def getForwards(forwardGetRequest: ForwardGetRequest, mailboxSession: MailboxSession): SFlux[ForwardGetResult] =
+    forwardGetRequest.ids match {
+      case None => getForwardsSingleton(mailboxSession)
+        .map(ForwardGetResult.found)
+        .flux()
+      case Some(ids) => SFlux.fromIterable(ids.value)
+        .flatMap(id => id match {
+          case UNPARSED_SINGLETON => getForwardsSingleton(mailboxSession)
+            .map(ForwardGetResult.found)
+          case _ => SMono.just(ForwardGetResult.notFound(id))
+        })
+    }
+
+  private def getForwardsSingleton(mailboxSession: MailboxSession): SMono[Forwards] = {
+    val userMailAddress: MailAddress = mailboxSession.getUser.asMailAddress
+    SMono.just(recipientRewriteTable.getStoredMappings(MappingSource.fromMailAddress(userMailAddress))
```

Comment:
It's green for the IT tests with memory, works as expected.

It's not for Distributed and I'm struggling.

When getting the mappings from `CassandraRecipientRewriteTable`, it does a `.blockOptional()` in the middle and reactor does not like it when I wrap it in my method processing... I tried to move the mappings out of the Mono but still get the same issue when calling `getStoredMappings`

The trace:
```
16:03:19.148 [ERROR] o.a.j.j.c.ProblemDetails - Unexpected error upon API request
java.lang.IllegalStateException: blockOptional() is blocking, which is not supported in thread reactor-http-epoll-2
	at reactor.core.publisher.BlockingOptionalMonoSubscriber.blockingGet(BlockingOptionalMonoSubscriber.java:104)
	at reactor.core.publisher.Mono.blockOptional(Mono.java:1752)
	at org.apache.james.rrt.cassandra.CassandraRecipientRewriteTable.getStoredMappings(CassandraRecipientRewriteTable.java:65)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.getForwardsSingleton(ForwardGetMethod.scala:128)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.getForwards(ForwardGetMethod.scala:115)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.doProcess(ForwardGetMethod.scala:87)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.doProcess(ForwardGetMethod.scala:77)
	at org.apache.james.jmap.method.MethodRequiringAccountId.$anonfun$process$2(Method.scala:61)
	at scala.util.Either.map(Either.scala:382)
	at org.apache.james.jmap.method.MethodRequiringAccountId.$anonfun$process$1(Method.scala:59)
	at scala.util.Either.flatMap(Either.scala:352)
	at org.apache.james.jmap.method.MethodRequiringAccountId.process(Method.scala:58)
	at org.apache.james.jmap.method.MethodRequiringAccountId.process$(Method.scala:56)
	at com.linagora.tmail.james.jmap.method.ForwardGetMethod.process(ForwardGetMethod.scala:77)
	at org.apache.james.jmap.routes.JMAPApi.$anonfun$processMethodWithMatchName$3(JmapApi.scala:89)
	at scala.util.Either.fold(Either.scala:189)
	at org.apache.james.jmap.routes.JMAPApi.$anonfun$processMethodWithMatchName$1(JmapApi.scala:89)
	at scala.Option.map(Option.scala:242)
	at org.apache.james.jmap.routes.JMAPApi.processMethodWithMatchName(JmapApi.scala:87)
	at org.apache.james.jmap.routes.JMAPApi.process(JmapApi.scala:81)
	at org.apache.james.jmap.routes.JMAPApi.$anonfun$processSequentiallyAndUpdateContext$5(JmapApi.scala:66)
	at reactor.core.scala.publisher.package$.$anonfun$scalaFunction2JavaFunction$1(package.scala:51)
	at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:152)
	at reactor.core.publisher.MonoFlatMapMany.subscribeOrReturn(MonoFlatMapMany.java:49)
	at reactor.core.publisher.FluxFromMonoOperator.subscribe(FluxFromMonoOperator.java:76)
	at reactor.core.publisher.FluxReplay.connect(FluxReplay.java:1120)
	at reactor.core.publisher.FluxAutoConnectFuseable.subscribe(FluxAutoConnectFuseable.java:63)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8469)
	at reactor.core.scala.publisher.SFlux.subscribe(SFlux.scala:909)
	at reactor.core.scala.publisher.SFlux.subscribe$(SFlux.scala:909)
	at reactor.core.scala.publisher.ReactiveSFlux.subscribe(ReactiveSFlux.scala:8)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.drain(FluxConcatMap.java:451)
	at reactor.core.publisher.FluxConcatMap$ConcatMapImmediate.onSubscribe(FluxConcatMap.java:219)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:165)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:87)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:127)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816)
	at reactor.core.publisher.MonoReduceSeed$ReduceSeedSubscriber.onComplete(MonoReduceSeed.java:165)
	at reactor.core.publisher.FluxIterable$IterableSubscription.fastPath(FluxIterable.java:362)
	at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:227)
	at reactor.core.publisher.MonoReduceSeed$ReduceSeedSubscriber.onSubscribe(MonoReduceSeed.java:113)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:165)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:87)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:157)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:119)
	at reactor.core.publisher.FluxHandle$HandleSubscriber.onNext(FluxHandle.java:119)
	at reactor.core.publisher.FluxMap$MapConditionalSubscriber.onNext(FluxMap.java:220)
	at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:130)
	at reactor.core.publisher.FluxHandleFuseable$HandleFuseableSubscriber.onNext(FluxHandleFuseable.java:184)
	at reactor.core.publisher.FluxContextWrite$ContextWriteSubscriber.onNext(FluxContextWrite.java:107)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1816)
	at reactor.core.publisher.MonoCollectList$MonoCollectListSubscriber.onComplete(MonoCollectList.java:128)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onComplete(FluxPeek.java:260)
	at reactor.core.publisher.FluxMap$MapSubscriber.onComplete(FluxMap.java:142)
	at reactor.netty.channel.FluxReceive.terminateReceiver(FluxReceive.java:468)
	at reactor.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:260)
	at reactor.netty.channel.FluxReceive.request(FluxReceive.java:129)
	at reactor.core.publisher.FluxMap$MapSubscriber.request(FluxMap.java:162)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:138)
	at reactor.core.publisher.MonoCollectList$MonoCollectListSubscriber.onSubscribe(MonoCollectList.java:79)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onSubscribe(FluxPeek.java:171)
	at reactor.core.publisher.FluxMap$MapSubscriber.onSubscribe(FluxMap.java:92)
	at reactor.netty.channel.FluxReceive.startReceiver(FluxReceive.java:167)
	at reactor.netty.channel.FluxReceive.lambda$subscribe$2(FluxReceive.java:146)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
```

Trying to investigate but not really sure what's going on there... or do we need to reactify that API?

Reply:
Nevermind... :)